### PR TITLE
[Merged by Bors] - chore(RingTheory): improve and generalize submodule localization API

### PR DESCRIPTION
**Pull request by alreadydone**
In Algebra/Module/LocalizedModule/Submodule, we introduce `Submodule.localized₀`, which is localization of a submodule considered as a submodule over the base ring rather than the localization. This allows us to talk about the localization of a submodule without choosing a specific ring localization, just like what `IsLocalizedModule` allows us to do for localization of a module.

As applications, `Rₚ` and every hypothesis depending on it are completely removed from the statement of `Submodule.le_of_localization_maximal`, etc. in RingTheory/LocalProperties/Submodule. We also reorder lemmas in this file to make the development more natural and golf the proofs (making use of `Module.eqIdeal` introduced in RingTheory/Ideal/Defs; also add some trivial lemmas.

In RingTheory/LocalProperties/Projective, we fix a proof due to removed explicit argument `Rₚ` and remove some unnecessary lines.

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Algebra/Module/LocalizedModule/Submodule.lean [hunk 1]
```diff
@@ -28,26 +28,36 @@ Results about localizations of submodules and quotient modules are provided in t
...
+  add_mem' := fun {x y} ⟨m, hm, s, hx⟩ ⟨n, hn, t, hy⟩ ↦ ⟨t • m + s • n, add_mem (M'.smul_mem t hm)
     (M'.smul_mem s hn), s * t, by rw [← hx, ← hy, IsLocalizedModule.mk'_add_mk']⟩
   zero_mem' := ⟨0, zero_mem _, 1, by simp⟩
-  smul_mem' := fun r x h ↦ by
-    have ⟨m, hm, s, hx⟩ := h
+  smul_mem' r x := by
+    rintro ⟨m, hm, s, hx⟩
+    exact ⟨r • m, smul_mem M' _ hm, s, by rw [IsLocalizedModule.mk'_smul, hx]⟩
+
+/-- Same as `Submodule.localized₀`, but viewed as an `S`-submodule. -/
```
Comment (erdOne) [inline Mathlib/Algebra/Module/LocalizedModule/Submodule.lean:50]: Can you copy the docstring of `Submodule.localized₀` here?

File: Mathlib/Algebra/Module/LocalizedModule/Submodule.lean [hunk 2]
```diff
@@ -28,26 +28,36 @@ Results about localizations of submodules and quotient modules are provided in t
...
-variable [CommRing R] [CommRing S] [AddCommGroup M] [AddCommGroup N]
+variable {R S M N : Type*}
+variable (S) [CommSemiring R] [CommSemiring S] [AddCommMonoid M] [AddCommMonoid N]
 variable [Module R M] [Module R N] [Algebra R S] [Module S N] [IsScalarTower R S N]
 variable (p : Submonoid R) [IsLocalization p S] (f : M →ₗ[R] N) [IsLocalizedModule p f]
 variable (M' : Submodule R M)
 
+section Submodule
+
 /-- Let `S` be the localization of `R` at `p` and `N` be the localization of `M` at `p`.
```
Comment (erdOne) [inline Mathlib/Algebra/Module/LocalizedModule/Submodule.lean:39]: There's no `S` at all in this declaration right?

**Review by erdOne — APPROVED** [review]
LGTM otherwise

Edit: other than the build error, that is

File: Mathlib/RingTheory/Ideal/Defs.lean [hunk 3]
```diff
@@ -68,6 +68,14 @@ theorem unit_mul_mem_iff_mem {x y : α} (hy : IsUnit y) : y * x ∈ I ↔ x ∈
 
 end Ideal
 
+/-- For two elements `m` and `m'` in an `R`-module `M`, the set of elements `r : R` with
+equal scalar product with `m` and `m'` is an ideal of `R`. -/
```
Comment (alreadydone) [inline Mathlib/RingTheory/Ideal/Defs.lean:72]:
```suggestion
equal scalar product with `m` and `m'` is an ideal of `R`. If `M` is a group, this coincides with the
kernel of `LinearMap.toSpanSingleton R M (m - m')`. -/
```

**Comment by Ruben-VandeVelde** [comment]
I feel like this would be much easier to review if it was split into more focused PRs. Probably each bullet point in your summary could be a separate PR.

**Comment by alreadydone** [comment]
[This](https://github.com/leanprover-community/mathlib4/pull/19118/files/9411282f586957e4a54470e4d352fe9cbc4f4950..1ef9d7321d796b51b23e80d7208361c5c8a82975) is the diff after Andrew's approval, and I agree the changes in [Mathlib/RingTheory/Ideal/Defs.lean](https://github.com/leanprover-community/mathlib4/pull/19118/files/9411282f586957e4a54470e4d352fe9cbc4f4950..1ef9d7321d796b51b23e80d7208361c5c8a82975#diff-b07d9879de8e5aaf8c99da0c8f04b1eed23daf73af82696409093926ee1f3afe) could be another PR (depending on this one). In [RingTheory/LocalProperties/Submodule.lean](https://github.com/leanprover-community/mathlib4/pull/19118/files/9411282f586957e4a54470e4d352fe9cbc4f4950..1ef9d7321d796b51b23e80d7208361c5c8a82975#diff-25704369c7e20d4992fa2ef4a09fc372db1df67c0e7c504d4a99820bb1ec8f46) (which is the main file I changed and Andrew reviewed) I merely reordered the lemmas to partially restore the original order, extracted `Module.eqIdeal` from two proofs, and added trivial lemmas about `= ⊤` and equality of linear maps.

**Comment by alreadydone** [comment]
@erdOne I hope this PR doesn't fall off your radar after I renamed the title :) If you think any part of it should be split, just let me know.

**Comment by jcommelin** [comment]
If it isn't too hard to split according to the bullet points in OP, I would appreciate that.

**Comment by alreadydone** [comment]
I discarded the part irrelevant to [#19080](https://github.com/leanprover-community/mathlib4/pull/19080#issuecomment-2480283549). This part has mostly been reviewed by @erdOne apart from a few later additions/refactors, and would be unnatural to break down further.

**Comment by jcommelin** [comment]
Thanks :tada:

bors merge

**Comment by alreadydone** [comment]
Forgot to mention: #19310 is the split-off part of this PR.

**Pull request closed without merging**